### PR TITLE
Add Support for Message Based Interrupt

### PR DESCRIPTION
**Pull request by afshin**
Implements handling of `interrupt_request` [message](https://jupyter-client.readthedocs.io/en/stable/messaging.html#kernel-interrupt), which is now possible due to the control channel being on a separate thread.
Does not implement handling on Windows, which is less likely to be used for remote kernels.

### CODE REVIEW

**Review by blink1073 — APPROVED** [review]
Excellent, thank you!

**Comment by blink1073** [comment]
cc @SylvainCorlay @kevin-bates

**Comment by SylvainCorlay** [comment]
Thanks for this!
Happy to see ipykernel move closer to fully implementing the protocol ;)

**Comment by ccordoba12** [comment]
I think this is not enough because it's missing a call to perform the request in jupyter_client, just like there's one for shutdown:

https://github.com/jupyter/jupyter_client/blob/9c0b4c0e28a9c17d9c14123522415b2048bd9522/jupyter_client/client.py#L762-L781

Otherwise, how is it possible to perform an interrupt from the kernel client?

**Comment by kevin-bates** [comment]
Hi @ccordoba12 - the interrupt machinery comes from the [`KernelManager`](https://github.com/jupyter/jupyter_client/blob/cbc29c9ae91b9a40c61b4b96937efaae4ea8c809/jupyter_client/manager.py#L536-L539) by virtue of the kernelspec indicating support for message-based interrupts.

Hmm - should we make signal-based interrupts the `else` so that kernelspecs that don't specify a mode attempt at least that form of interrupt?

**Comment by blink1073** [comment]
According to the [docs](https://jupyter-client.readthedocs.io/en/stable/messaging.html#kernel-interrupt), it is opt-in "For this to work, the kernels kernelspec must set interrupt_mode to message".  I think the question is whether the `ipykernel` default spec should be message based.

**Comment by kevin-bates** [comment]
ok - I understood the question to be one of missing logic - sorry about that.

My question was more general and I just found the answer in that ['signal' will be used if `interrupt_mode` is not specified](https://github.com/jupyter/jupyter_client/blob/cbc29c9ae91b9a40c61b4b96937efaae4ea8c809/jupyter_client/kernelspec.py#L38), which, now that I recall, have seen before. 

I'll let the maintainers decide what the desired mode should be. :smile:

**Comment by ccordoba12** [comment]
> the interrupt machinery comes from the KernelManager by virtue of the kernelspec indicating support for message-based interrupts.

Yeah, but if you connect to a kernel started elsewhere (for example, locally to a kernel in a server, which Spyder users do a lot), then you don't have access to a kernel manager, just a kernel client. And in that case it'd be really nice to be able to send an interrupt message using the control channel (like you can now to perform a shutdown with the method I referenced above).

I thought this PR was about that, but it seems I was mistaken, sorry.

**Comment by kevin-bates** [comment]
I'm sorry, I probably steered this discussion in the wrong direction and I apologize.  Please continue.

**Comment by ccordoba12** [comment]
No worries @kevin-bates. Let's ping @davidbrochart to see what he thinks about this one.

@davidbrochart, do you think it's worth it to add a message-based interrupt in jupyter-client?

**Comment by SylvainCorlay** [comment]
I think it is worth it just for the sake of completeness.

**Comment by ccordoba12** [comment]
Agreed. I'll open a PR in jupyter-client so that (hopefully) it can be part of its 7.0 version.

**Pull request merged**
